Implement fault tolerant GRIB2/PP loading.

ajdawson opened the pull request:
This PR allows loading of all convertible GRIB2 messages from a file even if some of the messages in the file cannot be converted to a cube. Since GRIB is inherently message based and there is not really a proper concept of a GRIB file, it seems unreasonable for Iris to fail completely when it meets a message it can't understand and fail to load the messages it does understand.

The implementation is relatively straightforward, an extra attribute `fault_tolerant` is added to the object `iris.fileformats.rules.Loader` which then allows `iris.fileformats.rules.load_cubes` to determine if it should try to carry on if it encounters a `TranslationError`. In principle this can be extended to any file format using `iris.fileformats.rules.load_cubes` as a handler, but only GRIB is implemented so far.

Remaining tasks:
- [ ] Push an integration test when the iris-test-data repository is updated with a suitable file.

marqh commented:
apologies for being late to the party

I support this change, I think it is a useful capability.  I would like to be clear that this is opt-in behaviour, I would expect to get exceptions from a load, and have to manually specify that I am happy to lose fields during load.

I think there should be a comment in the user guide section on loading data explaining this behaviour

marqh commented on lib/iris/fileformats/rules.py:
```diff
@@ -796,7 +799,14 @@ def load_cubes(filenames, user_callback, loader, filter_function=None):
+                                                         loader.converter)
+            except iris.exceptions.TranslationError as e:
+                if loader.fault_tolerant:
+                    skipped_fields += 1
```
I would like to see 'skipped_fields' as a container of error messages.  We can always get a length from a collection.  There is extras value in having the collection of error messages from the loader to help the user understand what has been skipped

I am not sure of a good way to make this available to the user, but if it could be done neatly then I'd be really pleased

rhattersley commented on lib/iris/fileformats/rules.py:
```diff
@@ -735,7 +735,8 @@ def _ensure_aligned(regrid_cache, src_cube, target_cube):
                                 ('field_generator', 'field_generator_kwargs',
-                                 'converter', 'legacy_custom_rules'))
+                                 'converter', 'legacy_custom_rules',
+                                 'fault_tolerant'))
```
As it stands, this is a backwards incompatible change. How much of an issue this is depends on whether it's targeting 1.9 or 2.0. (This kind of thing is why we might want to make more bits of the Iris API private.)

If we want to keep backwards compatibility then I think we need something like the following:
```python
class Loader(collections.namedtuple(...)):
    __slots__ = ()
    def __new__(cls, field_generator, ..., fault_tolerant=False):
        return tuple.__new__(cls, field_generator, ..., fault_tolerant)
```

pelson replied:
The alternative is that we just make the GRIB loader fault_tolerant globally. Provided the warning type is specific to GRIB loading, this can even be a case of:

```
warn(GRIBLoadFault, 'A specific bad thing happened')
```

And if we want to change the tolerance, we can change what happens when a GRIBLoadFault warning is encountered (e.g. make it raise an exception) through the standard warning mechanisms.

rhattersley commented:
> I would like to be clear that this is opt-in behaviour, I would expect to get exceptions from a load, and have to manually specify that I am happy to lose fields during load.

As long as translation errors result in visual feedback, I would expect opt-**out** to be more useful.

marqh commented:
Hello

I'd like to revitalise this PR

The comments have converged to an extent on this being a fault tolerant approach that just runs, then it is not backwards compatible.

If we can get this fixed and have slightly richer information on why a field is skipped I'm minded to get this in

So, is there life in this one? what do you think @ajdawson ?

ajdawson commented:
I'd like to get this in, but I'm not entirely clear on how I should proceed here. What are your requirements for information about skipped fields @marqh? Would we really want a separate warning for each field? Or a warning with a list of fields (containing what information?)?

I need to update this anyway to account for backwards compatibility in `Loader`. I'm not particularly keen on users having to go through the warnings system to generate exceptions, but I'm not sure what other options would be satisfactory.

ajdawson commented:
Just realised I already had a go at fixing this up, but hadn't pushed it up. I've rebased to catch up with master, and added a new commit that stores the exception from each skipped `TranslationError`. These messages are then raised as warnings of type `FieldLoadFault`. You can then use the warnings mechanism to control how many warnings you see, or turn them into exceptions (with `warnings.simplefilter('error', iris.exceptions.FieldLoadFault)`).

marqh commented:
many thanks for the updates @ajdawson 

please may you confirm that you think this is a backwards-compatible change, as discussed in 
https://github.com/SciTools/iris/pull/1703#discussion_r35311225
(it looks like it to me)

marqh commented on lib/iris/tests/integration/test_pp_constrained_load_cubes.py:
```diff
@@ -53,7 +55,8 @@ def test_pp_with_stash_constraints(self):
         filenames = [tests.get_data_path(('PP', 'globClim1', 'dec_subset.pp'))]
         pp_constraints = pp._convert_constraints(None)
-        pp_loader = iris.fileformats.rules.Loader(pp.load, {}, convert)
+        pp_loader = iris.fileformats.rules.Loader(pp.load, {}, convert, None,
```
I think this change is not required; are the defaults for the Loader class?

marqh commented on lib/iris/tests/integration/test_pp_constrained_load_cubes.py:
```diff
@@ -35,7 +35,8 @@ def test_pp_with_stash_constraint(self):
         pp_constraints = pp._convert_constraints(stcon)
-        pp_loader = iris.fileformats.rules.Loader(pp.load, {}, convert)
+        pp_loader = iris.fileformats.rules.Loader(pp.load, {}, convert, None,
+                                                  False)
```
I think this change is not required; are the defaults for the Loader class?

ajdawson replied:
That sounds correct, yes.

marqh commented on lib/iris/experimental/fieldsfile.py:
```diff
@@ -123,7 +123,8 @@ def load(filenames, callback=None):
     """
-    loader = Loader(_collations_from_filename, {}, _convert_collation, None)
+    loader = Loader(_collations_from_filename, {}, _convert_collation, None,
+                    False)
```
is this required?

ajdawson replied:
May be a left-over from old code.

marqh commented:
> I'd like to get this in, but I'm not entirely clear on how I should proceed here. What are your requirements for information about skipped fields @marqh? Would we really want a separate warning for each field? Or a warning with a list of fields (containing what information?)?

As long as there is enough information for a user to understand that they have lost something, I am happy

It would be good for me, as a user, to be able to browse the translation error reports
Does
https://github.com/SciTools/iris/pull/1703/files#diff-71a6ac4e2a835fb3910a8a8a8045474eR924
give me the report from the exception so I can get a clue about what I have lost?
Do i get enough information that I could run a 
grib_dump -O
on  the message and see what it looks like
Perhaps having the filename and the message number so I can go hunting is enough

I hope this gives a few clues on what I have in mind.  Are any of these fairly easy to implement?

mark

ajdawson commented:
If you turn warnings into exceptions, then it will fails at the first warning, which will be the first field, so no you would not be able to browse the full list of skipped fields. This sounds like a reasonable goal, but needs some thought on implementation.

An alternative would be one (possibly enormous) warning message detailing each skipped field, raising one warning would mean you'd get all the information if you turned warnings into exceptions. Other suggestions welcome.

rhattersley commented:
@marqh & I have been kicking around a slightly different idea:
```python
# Default behaviour just prints a single warning
>>> iris.load_cube(...)
__main__:1: FieldLoadFault

# Use a provided context manager to catch more detail.
# Detail could include filename, full exception/traceback, etc.
>>> with iris.fileformats.catch_faults() as faults:
...     iris.load_cube(...)
>>> print faults
[FieldLoadFault(filename='foo', message_id=3, error=<TranslationError....>),
FieldLoadFault(filename='foo', message_id=14, error=<TranslationError....>), ...]

# Can also use a context manager to state that no warnings are expected/wanted.
>>> with iris.fileformats.catch_faults(explode=True) as faults:
...     iris.load_cube(...)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TranslationError: ...
```

It doesn't seem possible to support this just by using warnings emitted from `iris.fileformats.rules.load_cubes()` (warnings don't preserve any non-standard attributes). But it does seem possible with slightly more custom handling within `load_cubes()`.

The whole thing could be controlled via a completely independent `FUTURE` flag (e.g. "soft_translation") and would work with *any* format that used `iris.fileformats.rules` (removing the need for the new `fault_tolerant` argument).

@ajdawson - your thoughts on this are very welcome!

ajdawson commented:
I like this idea, it makes for a nice simple way to handle the various cases. However, I've hit difficulties implementing it. I've got the whole structure working: a non-trivial subclass of warning with extra useful information attached, a context manager to catch the warnings if desired, all controlled by an `iris.FUTURE` option. The potential stumbling block I've hit is illustrated by this simple example:

```python
import warnings


def foo():
    print('the function ran')
    warnings.warn('and produced a warning!', UserWarning)


if __name__ == '__main__':
    foo()
    warnings.simplefilter('always', UserWarning)
    foo()
```

You might expect this to produce two warnings, but it doesn't on Python 2. I believe this is to do with http://bugs.python.org/issue4180, which was closed by application of a fix to Python 3.4+. Indeed the above works as expected on Python 3. This is a problem in the context of this work, in fact your example above is a use case that wouldn't actually work:

```python
# Default behaviour just prints a single warning
>>> iris.load_cube(...)
__main__:1: FieldLoadFault

# Use a provided context manager to catch more detail.
# Detail could include filename, full exception/traceback, etc.
>>> with iris.fileformats.catch_faults() as faults:
...     iris.load_cube(...)
>>> print faults
[]
```

Once the warning has been emitted once by the first load and registered, the filters are no longer checked by the warnings mechanism (it short circuits) so even though my implementation of `catch_faults()` sets the filter to "always" for `FieldLoadFault` I never get them again, so the list of reported faults would be empty even though there could be many faults...

ajdawson commented:
For reference: https://github.com/SciTools/iris/compare/master...ajdawson:fault-tolerant-load-wip.

rhattersley commented:
I don't think there's a (good) way to use warnings as the signal for all the different behaviours we want. Instead I'd add some explicit logic of our own and only raise warnings when they are wanted as the user-visible end result.

So after `iris.fileformats.rules.load_cubes()` has collected all the translation errors in `skipped_fields` perhaps something vaguely like:
```python
if skipped_fields:
    if _translation_mode == _WARNING:
        warnings.warn('{} fields could not be ...'.format(len(skipped_fields)), ...)
    else:
        _translation_errors.extend(skipped_fields)
```
... where `_translation_mode` and `_translation_errors` are globals (thread-local if we're feeling keen) which are controlled by the context manager. (The context manager wouldn't do anything with warnings at all.)

We could also refer to `_translation_mode` as we're catching the original translation errors and re-raise if `_translation_mode == _ERROR`.

ajdawson commented:
I don't think I'm going to be able to take this further for a good while. If someone else wants to take it forward in the mean time I have no objections.

marqh commented:
> I don't think I'm going to be able to take this further for a good while. If someone else wants to take it forward in the mean time I have no objections.

understood; I will try to get stuck in and see if I can implement the logic

thank you for all the efforts to date @ajdawson

pelson commented:
Closing as this has become stale.
